src/WebJobs.Script.WebHost/Management/LegionInstanceManager.cs

```diff
@@ -0,0 +1,146 @@
+    {
+        private static readonly object _assignmentLock = new object();
+        private static HostAssignmentContext _assignmentContext;
+
+        private readonly ILogger _logger;
+        private readonly IMetricsLogger _metricsLogger;
+        private readonly IMeshServiceClient _meshServiceClient;
+        private readonly IEnvironment _environment;
+        private readonly HttpClient _client;
+        private readonly IScriptWebHostEnvironment _webHostEnvironment;
+
+        public LegionInstanceManager(IHttpClientFactory httpClientFactory, IScriptWebHostEnvironment webHostEnvironment,
+            IEnvironment environment, ILogger<LegionInstanceManager> logger, IMetricsLogger metricsLogger, IMeshServiceClient meshServiceClient)
+        {
+            _client = httpClientFactory?.CreateClient() ?? throw new ArgumentNullException(nameof(httpClientFactory));
+            _webHostEnvironment = webHostEnvironment ?? throw new ArgumentNullException(nameof(webHostEnvironment));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _metricsLogger = metricsLogger ?? throw new ArgumentNullException(nameof(metricsLogger));
+            _meshServiceClient = meshServiceClient ?? throw new ArgumentNullException(nameof(meshServiceClient));
+            _environment = environment ?? throw new ArgumentNullException(nameof(environment));
+        }
+
+        public Task<string> SpecializeMSISidecar(HostAssignmentContext context)
+        {
+            // Legion will take care of MSI Specialization
+            return null;
+        }
+
+        public bool StartAssignment(HostAssignmentContext context)
+        {
+            if (!_webHostEnvironment.InStandbyMode && !context.Environment.TryGetValue(EnvironmentSettingNames.ContainerStartContext, out string startContext))
+            {
+                // This is only true when specializing pinned containers.
+                _logger.LogError("Assign called while host is not in placeholder mode and start context is not present.");
+                return false;
+            }
+
+            if (_environment.IsContainerReady())
+            {
+                _logger.LogError("Assign called while container is marked as specialized.");
+                return false;
+            }
+
+            if (context.IsWarmupRequest)
+            {
+                // No need to download since file already exists
+                return true;
+            }
+
+            if (_assignmentContext == null)
+            {
+                lock (_assignmentLock)
+                {
+                    if (_assignmentContext != null)
+                    {
+                        return _assignmentContext.Equals(context);
+                    }
+                    _assignmentContext = context;
+                }
+
+                _logger.LogInformation($"Starting Assignment. Cloud Name: {_environment.GetCloudName()}");
```

Comment:
structured logged here too + any logging statements
```suggestion
                _logger.LogInformation("Starting Assignment. Cloud Name: {name}", _environment.GetCloudName());
```

Reply:
+1 - but I would recommend `{cloudName}` over `{name}`, as you need to consider what other log scopes might be active. "name" is generic and may conflict.

Reply:
Fixed!